nixos/modules/services/matrix/meowlnir.nix

```diff
@@ -0,0 +1,309 @@
+        and then specify `MEOWLNIR_APPSERVICE_AS_TOKEN="{token}"` in the environment file.
+        This value will get substituted into the configuration file as as token.
+      '';
+    };
+
+    serviceDependencies = lib.mkOption {
+      type = with lib.types; listOf str;
+      default = lib.optional config.services.matrix-synapse.enable config.services.matrix-synapse.serviceUnit;
+      defaultText = lib.literalExpression ''
+        optional config.services.matrix-synapse.enable config.services.matrix-synapse.serviceUnits
+      '';
+      description = ''
+        List of Systemd services to require and wait for when starting the application service.
+      '';
+    };
+
+    registerToSynapse = lib.mkOption {
+      type = lib.types.bool;
+      default = config.services.matrix-synapse.enable;
+      defaultText = lib.literalExpression "config.services.matrix-synapse.enable";
+      description = ''
+        Whether to add the bridge's app service registration file to
+        `services.matrix-synapse.settings.app_service_config_files` and enable
+        the required experimental features.
+      '';
+    };
+  };
+
+  config = lib.mkIf cfg.enable {
+    users = {
+      users.meowlnir = {
+        home = cfg.dataDir;
+        description = "Meowlnir bridge user";
+        group = "meowlnir";
+        isSystemUser = true;
+      };
+      groups.meowlnir = { };
+    };
+
+    services.matrix-synapse = lib.mkIf cfg.registerToSynapse {
+      settings = {
+        app_service_config_files = [ registrationFile ];
+        experimental_features = {
+          # Actually MSC4203, but it was previously a part of MSC2409
+          msc2409_to_device_messages_enabled = true;
+          # MSC3202 has two parts, both need to be enabled
+          msc3202_device_masquerading = true;
+          msc3202_transaction_extensions = true;
+        };
+      };
+    };
+    systemd.services.matrix-synapse = lib.mkIf cfg.registerToSynapse {
+      serviceConfig.SupplementaryGroups = [ "meowlnir" ];
+    };
+
+    services.meowlnir.settings = {
+      # Note: this is defined here to avoid the docs depending on `config`
+      homeserver.domain = lib.mkDefault config.services.matrix-synapse.settings.server_name;
+    };
+
+    systemd.services.meowlnir = {
+      description = "meowlnir - opinionated Matrix moderation bot";
+
+      wantedBy = [ "multi-user.target" ];
+      wants = [ "network-online.target" ] ++ cfg.serviceDependencies;
+      after = [ "network-online.target" ] ++ cfg.serviceDependencies;
+
+      restartTriggers = [ settingsFileUnsubstituted ];
+      preStart = ''
+        # ensure that the data directory is set up correctly
+        mkdir -p ${cfg.dataDir}
+        chmod 755 ${cfg.dataDir}
+
+        # substitute the settings file by environment variables
+        # in this case read from EnvironmentFile
+        test -f '${settingsFile}' && rm -f '${settingsFile}'
+        old_umask=$(umask)
```

Comment:
I think it would be a little easier to do this with a subshell, i.e.

```
(
  umask 0177
  ...
)
```

Then you don't have to care about restoring this.